8296886: Fix various include sort order issues

stefank opened the pull request:
The sorted blocks of includes have deteriorated to the point that I felt compelled to clean up some of the issues.

*EDIT*: The below discussion has been deferred out of this PR. Now this only deals with fixing the placement and sorting of includes, plus some surrounding blank lines.

One of the more prevalent issues is that files in src/hotspot/share/include are not properly sorted. There has been some discussion that that was done on purpose, but it just adds another exception to the include rules that don't have any practical purposes, IMHO. It also goes against our written style guide around include files. One argument why it was OK have the files in include/ pushed up to the top of the sorted block, was that the file was included without specifying a directory. That's an argument that contradicts how we treat platform-dependent files, which (unfortunately) often also are specified without a prefixed directory, so I don't think that's a good enough argument, again IMHO. To remove this special case, I've removed the extraneous make file entry to have src/hotspot/share/include in the set of directories to search for headers when compiling HotSpot. Now all the header files in src/hotspot/share/include gets included by specifying the path from src/hotspot/share, just like the other platform-independent headers in HotSpot.

While going over the include headers I've also cleaned up surrounding whitespaces and incorrect include guards.

<!-- Anything below this marker will be automatically updated, please do not edit manually! -->
---------
### Progress
- [x] Change must be properly reviewed (1 review required, with at least 1 [Reviewer](https://openjdk.org/bylaws#reviewer))
- [x] Change must not contain extraneous whitespace
- [x] Commit message must refer to an issue

### Issue
 * [JDK-8296886](https://bugs.openjdk.org/browse/JDK-8296886): Fix various include sort order issues


### Reviewers
 * [Kim Barrett](https://openjdk.org/census#kbarrett) (@kimbarrett - **Reviewer**) ⚠️ Review applies to [e15527d3](https://git.openjdk.org/jdk/pull/11108/files/e15527d383cb8fe953cfe3f1bac8a45f3c3557cb)
 * [David Holmes](https://openjdk.org/census#dholmes) (@dholmes-ora - **Reviewer**) ⚠️ Review applies to [e15527d3](https://git.openjdk.org/jdk/pull/11108/files/e15527d383cb8fe953cfe3f1bac8a45f3c3557cb)
 * [Thomas Stuefe](https://openjdk.org/census#stuefe) (@tstuefe - **Reviewer**) ⚠️ Review applies to [e15527d3](https://git.openjdk.org/jdk/pull/11108/files/e15527d383cb8fe953cfe3f1bac8a45f3c3557cb)


### Reviewing
<details><summary>Using <code>git</code></summary>

Checkout this PR locally: \
`$ git fetch https://git.openjdk.org/jdk pull/11108/head:pull/11108` \
`$ git checkout pull/11108`

Update a local copy of the PR: \
`$ git checkout pull/11108` \
`$ git pull https://git.openjdk.org/jdk pull/11108/head`

</details>
<details><summary>Using Skara CLI tools</summary>

Checkout this PR locally: \
`$ git pr checkout 11108`

View PR using the GUI difftool: \
`$ git pr show -t 11108`

</details>
<details><summary>Using diff file</summary>

Download this PR as a diff file: \
<a href="https://git.openjdk.org/jdk/pull/11108.diff">https://git.openjdk.org/jdk/pull/11108.diff</a>

</details>

fisk commented:
Hmm. Now that "jvm.hpp" becomes "include/jvm.hpp" it seems like it would make sense if the include guard for jvm.hpp reflected the relative file path, like most other files. I noticed that its current include guard is `#ifndef _JAVASOFT_JVM_H_`. When I search for "javasoft" I get no hits. I guess it isn't obvious if we can change it without introducing build issues for users out there? So maybe it should remain the way it is anyway.

kimbarrett commented on src/hotspot/share/prims/scopedMemoryAccess.cpp:
```diff
@@ -23,18 +23,17 @@
+#include "prims/stackwalk.hpp"
 #include "runtime/interfaceSupport.inline.hpp"
+#include "runtime/jniHandles.inline.hpp"
+#include "runtime/deoptimization.hpp"
```
runtime/deoptimization.hpp should be at the front of the runtime/ list.

kimbarrett commented on src/hotspot/share/prims/whitebox.cpp:
```diff
@@ -23,8 +23,6 @@
  */
 
 #include "precompiled.hpp"
-#include <new>
```
Why was `<new>` removed?

stefank replied:
We tend to pull it in via memory/allocation.hpp. I can revert this, but I'm not sure it is important. There's no consistency in when we include `<new>` or not.

kimbarrett commented on test/hotspot/gtest/jfr/test_adaptiveSampler.cpp:
```diff
@@ -47,6 +45,8 @@
 
 #include "unittest.hpp"
 
+#include <cmath>
```
Why is this after unittest.hpp?

stefank replied:
System includes go after HotSpot includes.

kimbarrett commented on test/hotspot/gtest/metaprogramming/test_enableIf.cpp:
```diff
@@ -26,9 +26,11 @@
+
 #include "unittest.hpp"
 
+#include <type_traits>
```
Why is this following unittest.hpp.

kimbarrett commented on test/hotspot/gtest/runtime/test_os_linux.cpp:
```diff
@@ -25,16 +25,17 @@
 
-#include <sys/mman.h>
-
+#include "os_linux.hpp"
```
Why do we even need this - we have runtime/os.hpp and we're on Linux.

kimbarrett commented on test/hotspot/gtest/utilities/test_align.cpp:
```diff
@@ -25,9 +25,11 @@
+
 #include "unittest.hpp"
 
+#include <limits>
```
Why after unittest.hpp?

kimbarrett commented on test/hotspot/gtest/utilities/test_bitMap_setops.cpp:
```diff
@@ -22,15 +22,17 @@
+
 #include "unittest.hpp"
 
+#include <stdlib.h>
```
Why after unittest.hpp.

kimbarrett commented on test/hotspot/gtest/utilities/test_count_leading_zeros.cpp:
```diff
@@ -26,6 +26,7 @@
+
 #include "unittest.hpp"
 
 #include <limits>
```
[pre-existing] Why after unittest.hpp?

kimbarrett commented on test/hotspot/gtest/utilities/test_enumIterator.cpp:
```diff
@@ -23,9 +23,11 @@
+
 #include "unittest.hpp"
 
+#include <type_traits>
```
Why after unittest.hpp?

kimbarrett commented on test/hotspot/gtest/utilities/test_globalDefinitions.cpp:
```diff
@@ -27,9 +27,11 @@
+
 #include "unittest.hpp"
 
+#include <type_traits>
```
Why after unittest.hpp.

kimbarrett commented on test/hotspot/gtest/utilities/test_nonblockingQueue.cpp:
```diff
@@ -27,9 +27,9 @@
+
 #include "threadHelper.inline.hpp"
 #include "unittest.hpp"
-#include <new>
```
Why removing `<new>`?  (and pre-existing, why after unittest.hpp?)

kimbarrett commented on test/hotspot/gtest/utilities/test_population_count.cpp:
```diff
@@ -27,9 +27,11 @@
+
 #include "unittest.hpp"
 
+#include <limits>
```
Why after unittest.hpp?

kimbarrett commented on test/hotspot/gtest/utilities/test_powerOfTwo.cpp:
```diff
@@ -25,9 +25,11 @@
+
+#include "unittest.hpp"
+
 #include <limits>
```
Why after unittest.hpp?

kimbarrett reviewed (changes requested):
I don't think platform-specific files that are included without providing a
directory should be sorted with the shared files on just the filename.  That
just scatters the platform-specific files, in a way that I think isn't
helpful.  Such files ought to be at the end, using the appropriate macros.
But there are some filenames that don't have the appropriate stem for use with
those macros.  I think those files should be renamed.  So I disagree with
moving the includes of these files.  I'd prefer they were left as-is by this
change, with a followup to rename files and use the appropriate macros.

There are some files external to hotspot that are being included without any
directory component, such as jni.h and jimage.hpp. Moving them into the middle
of the "shared" includes, sorted by just filename name without a directory,
seems like it is hiding external dependencies. I don't have a well-formed
suggestion for what to do about them, but I think the proposed moves aren't
helpful.

I wonder if standard library includes should also be sorted.  I think you
might have done that in some places, but not sure about that.  Certainly there
are some that aren't.

There are a number of gtest where a standard library include was moved to
after the include of unittest.hpp. I think unittest.hpp is supposed to always
be the last included header, and added a comment to those places. But now I
can't find anything that says that. (BTW, I agree with moving things like
threadHelper.inline.hpp into a block after normal includes, with
unittest.hpp.)

Related, there are 62 .cpp files missing "precompiled.hpp"
`find . -name "*.cpp" -exec grep -L "precompiled.hpp" {} \;`
At least some (os_linux.cpp, os_posix.cpp) have a comment claiming
intentionality, though that seems suspect (os_linux.cpp has been that way
since before mercurial). I think this should be another followup.

I wish this change had been split up a bit. One change (or a separate commit
within this change) that dealt with files in share/include could have been
mechanically checked or pretty easily verfied by eye. That would have left a
much smaller residue that needed going through more carefully.

Touching a whole mess of gtests to put a blank line between other includes and
the final include of unittest.hpp could also have been separate (PR or commit).

Should there be any additions to the style guide to codify some of these
changes?  I can't find anything about where unittest.hpp and other
gtest-specific headers should go.  Nor can I find anything about where
standard library headers should go.

stefank commented:
I'll split this out into separate changes.

* Regarding in `#include "include/xxx.h"`: This used to be `#include "prims/xxx.h"` and sorted the rest of the header file. Then they got moved to include/ and the directory was dropped for some reason. @dholmes-ora maybe you have a better name than `include/`, that is less eye jarring?

* Remember, a lot of the order we used was generated by a script. Some of the complaints above are against the style that was used ever since we added these includes to the files.

* System files comes after the HotSpot includes, with a blank line before.

* Platform files don't necessarily use the macro stems and then don't sort after the shared includes. I'm fine with changing this, but that would have to be tackled as a separate proposal.

* Sounds like a good proposal to move the external, shared headers like jimage.h and jni.h to a separate section *after* the HotSpot includes.

* I also remember that unittest.h needs to go at the end of the list, and have seen many violations of that. I added a blank line as a suggestion that this might give a visual aid of what's going on.

* <new> is typically provided via allocation.hpp, but I can leave the few odd <new> includes.

stefank commented:
Closing this down and restarting this discussion with limited patch, which only adds the missing include/ dir: #11133.

kimbarrett commented on src/hotspot/os/windows/jvm_windows.cpp:
```diff
@@ -23,9 +23,9 @@
-#include "jvm.h"
-#include "runtime/interfaceSupport.inline.hpp"
+#include "include/jvm.h"
 #include "os_windows.hpp"
```
But should we be directly including os_windows.hpp, rather than including os.hpp?

stefank replied:
Deferring those discussions to a potential separate RFE.

stefank commented on src/hotspot/cpu/x86/macroAssembler_x86.cpp:
```diff
@@ -23,15 +23,16 @@
 #include "asm/assembler.inline.hpp"
 #include "compiler/compiler_globals.hpp"
 #include "compiler/disassembler.hpp"
+#include "crc32c.h"
```
I'd like to defer that discussion to a separate RFE.

stefank commented on src/hotspot/os_cpu/bsd_zero/os_bsd_zero.cpp:
```diff
@@ -57,6 +52,11 @@
 
+#if !defined(__APPLE__) && !defined(__NetBSD__)
+#include <pthread.h>
+# include <pthread_np.h> /* For pthread_attr_get_np */
```
I reverted all such space cleanups, in the interest of getting the limited cleanup fixed. If we want to fix these spaces I'd like that to happen as a separate RFE.

stefank commented on src/hotspot/share/cds/classListParser.cpp:
```diff
@@ -37,9 +35,11 @@
 #include "interpreter/bytecode.hpp"
 #include "interpreter/bytecodeStream.hpp"
 #include "interpreter/linkResolver.hpp"
+#include "jimage.hpp"
```
We can hopefully discuss this as a separate RFE.

dholmes-ora reviewed (approved):
Seems okay and generally uncontroversial :)

I'm not sure why conditional includes (that don't rely on macros.hpp) need to come at the end rather than in normal sort order? I don't care either way but a rationale for this would be good if it is to be the preferred style.

Thanks.

kimbarrett commented:
> I'm not sure why conditional includes (that don't rely on macros.hpp) need to come at the end rather than in normal sort order? I don't care either way but a rationale for this would be good if it is to be the preferred style.

Because the Style Guide says:
* Put conditional inclusions (`#if ...`) at the end of the include list.

I think most of our conditional includes these days are to support conditional features.  It makes sense to group all the additional includes related to a feature.

kimbarrett reviewed (approved):
Looks good.

dholmes-ora commented:
> > I'm not sure why conditional includes (that don't rely on macros.hpp) need to come at the end rather than in normal sort order? I don't care either way but a rationale for this would be good if it is to be the preferred style.
> 
> Because the Style Guide says:
> 
>     * Put conditional inclusions (`#if ...`) at the end of the include list.

Ah I see. Thanks for that @kimbarrett  
> 
> I think most of our conditional includes these days are to support conditional features. It makes sense to group all the additional includes related to a feature.

True. There are a lot of single includes but it makes sense to have one simple rule.

tstuefe reviewed (approved):
Looks good and is an improvement. About the removed comments after includes, I am sometimes guilty of this to mark includes that came just for one specific little thing - the implied hope is that disentangling includes and removing unnecessary ones periodically is easier that way. However, this never worked. I wish we had an automatic include "GC" process to reduce the includes, like Ioi sometimes does manually.

stefank commented:
Thanks all for reviewing! I'm going to let the GHA run before integrating this change.

stefank commented:
/integrate
